### ISSUE TITLE
Does Randomized need to use a List?

### ISSUE THREAD

**Issue by StuporHero** [yegor256/cactoos#1427]
[Randomized](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/text/Randomized.java) only takes a `List<Character>` or a `Character...` for the list of characters from which to randomly draw. None of the public methods of the `List` interface are used in the class, so I'd like to request it be changed to accept `Iterable<Character>` instead. If it must use `List` (e.g. to guarantee order for the characters), a note about that should be added to the Javadoc comment on the class so others like myself find the answer in the code.

Alternatively, having it accept a `Scalar<List<Character>>` would allow me to avoid implementing the rather obnoxious `List` interface to make standardized character lists. Also, some of the constructors have `len` first, and some have `chrs` first which should probably be standardized.

Tell you what. Let me know if it has to be `List`, and I'll fix it. The constructors are all over the place, and it would be easier to just clean it up and discuss it in a PR.

Thank you in advance.

**Comment by StuporHero** [comment]
@victornoel No, thanks. I have too much going on to receive tasks that aren't directly related to what I'm working on. If that changes, though, I'll let you know. I'm already a part of Zerocracy, so it would be a straightforward addition.

**Comment by victornoel** [comment]
@StuporHero ok, and what do you think about my answer to the ticket? It seems `List` is needed so that `get` can be called on it.

**Comment by StuporHero** [comment]
@victornoel I must have missed that. In that case, adding a constructor that accepts `Scalar<List<Character>>` would be nice so I can implement that interface instead of `List<Character>` to create my own custom character classes to feed into it. I want to use it to create a randomized component for the names of database objects, but I have several particular constraints that would be easier to work with as dedicated classes rather than repeating `new ListOf<>(...)`. Having a constructor accept that would allow me to instantiate Randomized in a constructor while avoiding a call to the `value()` method.

**Comment by victornoel** [comment]
@StuporHero it's not completely clear to me, could you show an example of usage of this new constructor you propose? I would like to get a better idea of the needs before jumping into a solution and it would help better design the rest of the lib :)

**Comment by StuporHero** [comment]
@victornoel Let's say I have a particular set of characters available to me for a name, and I want part of the name to be `Randomized` from that set. I want to be able to define that set of characters once in a class and use instances of that class as arguments in the constructor for `Randomized`.

```java
public final class MyClass {
    private final Text name;

    public MyClass(final String name) {
        this.name = new FormattedText(
            "%s-%s",
            name,
            new Randomized(
                5,
                new LimitedCharacterSet()
            )
        );
    }
}
```

Right now, in order for the above code to work, `LimitedCharacterSet` has to implement `List<Character>`, which is an inconvenient interface to have to implement. If `LimitedCharacterSet` implemented `Scalar<List<Character>>`, I would have to call its `value()` method which bars us from using it directly in the constructor of another class unless we use `Scalar<Randomized>`, but using `Scalar<Randomized>` prevents us from using it as a part of a `FormattedText` without using `Scalar<FormattedText>`.

```java
public final class MyClass {
    private final Scalar<? extends Text> name;

    public MyClass(final String name) {
        this.name = () -> new FormattedText(
            "%s-%s",
            name,
            new Randomized(
                5,
                new LimitedCharacterSet().value()
            )
        );
    }
}
```

 While the code isn't horrible, it's not quite as easy to comprehend as if I could just pass a `Scalar<List<Character>>` to the constructor for `Randomized`.

(This makes me think perhaps `PaddedStart` could could get a `PaddedEnd` sibling, and they could both be adapted to use `Randomized` to accomplish what I'm trying to do, but that would be another ticket, obviously).

**Comment by victornoel** [comment]
@StuporHero and could you show me your ideal declaration for `LimitedCharacterSet`? It implements `Set<Character>`?

**Comment by StuporHero** [comment]
@victornoel 

> Alternatively, having it accept a `Scalar<List<Character>>` would allow me to avoid implementing the rather obnoxious `List` interface.

Now that you mention it, `List` really should be `Set`, but `Set` also lacks the `get` method. Since the use of `get` is more of an implementation detail than a proper constraint on acceptable inputs (element uniqueness is meaningful; order is not), perhaps it should accept a `Set` and either wrap it in `ListOf` for access to `get` or use `toArray()` and access the elements that way. I'm also realizing that the abstract `*Envelope` classes are specifically designed to meet my needs rather than requiring `Randomized` to accept a `Scalar`.

```java
public final class LimitedCharacterSet extends SetEnvelope<Character> {
    public LimitedCharacterSet() {
        super(
            new SetOf<>(
                'A', 'B', 'C',
                '1', '2', '3'
            )
        );
    }
}
```

Sorry that this has been a bit of a journey, but after all this, I'd like `Randomized` to accept `Set<Character>` instead of `List<Character>` because it is more semantically accurate and for the parameter order between the constructors to be made consistent.

**Comment by victornoel** [comment]
@StuporHero let's be careful not to mixup problem and solution :) 

For example, would you be ok with?
```java
new Randomized(
  5,
  new ListOf<>(new LimitedCharacterSet())
)
```

Also, yes, `Randomized` could take an `Iterable` and we create internally a `List` from it in order to have get, that's a possible solution, so that you can directly pass any `Iterable` as you originally asked.

I believe that the constructor takes a `List` because this allows to deterministically specify the order of the characters passed to the class. Let's not forget that the constructor of a class is part of its implementation, not its interface!

**Comment by StuporHero** [comment]
@victornoel Since the `Character`s are chosen at random, I don't see why it would be a requirement that an incoming collection have a deterministic order as a rule for this class. If a given use-case requires a deterministic collection, a `SortedSet` could be used. However, if an incoming collection has more than one of the same element, which is possible if we use `Collection`, `Iterable`, or `List`, the probability that a given `Character` will be selected changes relative to the total number of `Character`s in the collection. If we enforce a `Set`, we enforce an equal probability of any particular `Character` being chosen which seems to me like a desirable trait for this particular class. What do you think?

**Comment by victornoel** [comment]
@StuporHero 
> If we enforce a `Set`, we enforce an equal probability of any particular `Character` being chosen which should maximize randomness. This seems to me like a desirable trait for this particular class. What do you think?

Actually one would maybe want that some characters have more probability to be chosen.

Anyway, this is not the discussion here, you didn't tell me if my proposal was acceptable, if not, let's introduce a secondary constructor taking an `Iterable` that calls new `ListOf<>(iterable)`.

**Comment by StuporHero** [comment]
@victornoel I think we lost the plot on this ticket once I realized I could just extend `ListEnvelope` to solve my original problem (which was that I found the `List` interface cumbersome to implement as it relates to creating standardized `List`s of `Character`s to use in my own libraries). Knowing that, the original reason for me opening this ticket no longer applies. I would, however, like to discuss the subject of which types of collection a `Randomized` should take. Should I close this ticket and open another one? I have some other thoughts on the subject that might be of interest.
